File: solutions/c/protein-translation/1/protein_translation.c
#include "protein_translation.h"
#include <string.h>

proteins_t proteins(const char *const rna){
    proteins_t result;
    result.count = 0;
    result.valid = true;
    int len = strlen(rna);
    if(len == 0) return result;
    int nCodons = len/3;
    if(nCodons <= 0){
        result.valid = false;
        return result;
    }
    for(int i = 0; i < nCodons; i++){
        char codon[4];
        codon[0] = rna[(i*3)];
        codon[1] = rna[(i*3)+1];
        codon[2] = rna[(i*3)+2];
        codon[3] = '\0';
        if(strcmp(codon, "AUG") == 0){
            result.proteins[i] = Methionine;
        }else if(strcmp(codon, "UUU") == 0 || strcmp(codon, "UUC") == 0){
            result.proteins[i] = Phenylalanine;
        }else if(strcmp(codon, "UUA") == 0 || strcmp(codon, "UUG") == 0){
            result.proteins[i] = Leucine;
        }else if(strcmp(codon, "UCU") == 0 || strcmp(codon, "UCC") == 0 || strcmp(codon, "UCA") == 0 || strcmp(codon, "UCG") == 0){
            result.proteins[i] = Serine;
        }else if(strcmp(codon, "UAU") == 0 || strcmp(codon, "UAC") == 0){
            result.proteins[i] = Tyrosine;
        }else if(strcmp(codon, "UGU") == 0 || strcmp(codon, "UGC") == 0){
            result.proteins[i] = Cysteine;
        }else if(strcmp(codon, "UGG") == 0){
            result.proteins[i] = Tryptophan;
        }else if(strcmp(codon, "UAA") == 0 || strcmp(codon, "UAG") == 0 || strcmp(codon, "UGA") == 0){
            return result;
        }
        else result.valid = false;
        result.count++;
    }
    if(rna[nCodons*3] != '\0') result.valid = false;
    return result;
}

/*
ERRORES:
    NO HAY ERRORES
*/

File: solutions/zig/protein-translation/1/protein_translation.zig
const std = @import("std");
const mem = std.mem;

pub const TranslationError = error{
    InvalidCodon,
};

pub const Protein = enum {
    methionine,
    phenylalanine,
    leucine,
    serine,
    tyrosine,
    cysteine,
    tryptophan,
    stop,
};

pub fn proteins(allocator: mem.Allocator, strand: []const u8) (mem.Allocator.Error || TranslationError)![]Protein {
    const len = strand.len;
    var result = try std.ArrayList(Protein).initCapacity(allocator, 64);
    errdefer result.deinit(allocator);

    var i: usize = 0;
    while (i + 3 <= len) : (i += 3) {
        const codon = strand[i .. i + 3];
        const protein = try codonToProtein(codon);
        if (protein == .stop) return result.toOwnedSlice(allocator);
        try result.append(allocator, protein);
    }
    if (i != strand.len) return TranslationError.InvalidCodon;
    return result.toOwnedSlice(allocator);
}

fn codonToProtein(codon: []const u8) TranslationError!Protein {
    if (codon.len != 3) return TranslationError.InvalidCodon;
    if (std.mem.eql(u8, codon, "AUG")) {
        return .methionine;
    } else if (std.mem.eql(u8, codon, "UUU") or std.mem.eql(u8, codon, "UUC")) {
        return .phenylalanine;
    } else if (std.mem.eql(u8, codon, "UUA") or std.mem.eql(u8, codon, "UUG")) {
        return .leucine;
    } else if (std.mem.eql(u8, codon, "UCU") or std.mem.eql(u8, codon, "UCC") or std.mem.eql(u8, codon, "UCA") or std.mem.eql(u8, codon, "UCG")) {
        return .serine;
    } else if (std.mem.eql(u8, codon, "UAU") or std.mem.eql(u8, codon, "UAC")) {
        return .tyrosine;
    } else if (std.mem.eql(u8, codon, "UGU") or std.mem.eql(u8, codon, "UGC")) {
        return .cysteine;
    } else if (std.mem.eql(u8, codon, "UGG")) {
        return .tryptophan;
    } else if (std.mem.eql(u8, codon, "UAA") or std.mem.eql(u8, codon, "UAG") or std.mem.eql(u8, codon, "UGA")) {
        return .stop;
    }
    return TranslationError.InvalidCodon;
}
